Add counting mode, and only programming languages flags. (#110)

zkry opened the pull request:
#### What the commit does
This commit addresses some of the problems in issue #110 to address compatibility with linguist. To do this two main flags were added: one to choose the way the files are counted, and another to exclude the program from counting non-programming languages. When working on the line-counting mode, I noticed a bug in the way non-blank lines were counted where two or more blank lines in a row would throw the count off. It has a test too.

#### Things to look out for
Let me know if there is anything I can improve on the documentation of these flags. Also, do you have any recommendations how I can get the two programs to analize the same files. linguist seems to skip over certain files and I'm not sure why.

juanjux commented:
Hi zkry,

Sorry for waiting so much to reply to you, as a poor excuse we've had some maintainership changes on this project so your PR got in review limbo.

It LGTM, but travis is failing, I'll check the problem in a few hours.

dennwc commented on cli/enry/main.go:
```diff
@@ -197,10 +278,30 @@ func printFileAnalysis(file string) {
-	totalLines := strings.Count(content, "\n")
-	nonBlank := totalLines - strings.Count(content, "\n\n")
-	return totalLines, nonBlank
+func getLines(content []byte) (total int, nonBlank int) {
```
It will be a bit nicer to use `bufio.Scanner` here because it will not require loading the whole file into memory and the code for blank line detection would be simpler:
```golang
line := scanner.Bytes()
line = bytes.TrimSpace(line)
isBlank := len(line) == 0
```

dennwc commented on cli/enry/main.go:
```diff
@@ -159,21 +167,94 @@ func printJson(out map[string][]string, buff *bytes.Buffer) {
+	var filesErr filelistError
+	var t float64
+	for _, fName := range files {
+		f, err := os.Open(fName)
```
You are leaking file descriptors here. Close file directly after `Stat`, but before the `if` statement.

zkry commented:
No problems :) I addressed the issues brought up. The only thing I wasn't sure about was avoiding loading the file all at once. For the getLines function, I did the recommended change and it made the function much simpler but since much of the other enry functions accept []byte, I had a hard time figuring out how to avoid loading the file all at once.

dennwc commented:
@zkry Commit looks valid, and the second bot is okay with it. Squash looks good, but it looks like the rebase was on top of old master - Github still shows the conflict. Please update the master first before rebasing, or fetch and rebase onto `origin/master`.

@mcuadros Can you please check what's wrong with the bot? It's the second valid PR it rejects.

zkry commented:
Ok, I've got it now. I was confused because I had gopkg.io as my origin remote. Thanks

dennwc commented:
Looks good now! Bots are also happy, but CI fails:
```
# gopkg.in/src-d/enry.v1/cli/enry
../../../gopkg.in/src-d/enry.v1/cli/enry/main_test.go:30:28: undefined: getLines
```

dennwc commented:
Thanks again for contribution and sorry for these conflicts :)

zkry commented:
And thanks for helping me work through this :)
